pkg/ring/incremental_transfer.go

```diff
@@ -0,0 +1,391 @@
+				Op:            op,
+				LeftInclusive: true,
+				MaxHeartbeat:  i.cfg.RingConfig.HeartbeatTimeout,
+			}
+			rhsRange := RangeOptions{
+				Range: TokenRange{From: token.Token, To: target.Token},
+
+				ID:             i.ID,
+				Op:             op,
+				RightInclusive: true,
+				MaxHeartbeat:   i.cfg.RingConfig.HeartbeatTimeout,
+			}
+
+			if d.InRange(lhsRange) || d.InRange(rhsRange) {
+				continue
+			}
+
+			addr := d.Ingesters[target.Ingester].Addr
+			ret[addr] = append(ret[addr], TokenRange{
+				From: startRange.Token,
+				To:   endRange.Token,
+			})
+		}
+	}
+
+	if lastErr != nil {
+		level.Error(util.Logger).Log(
+			"msg", fmt.Sprintf("failed to find complete transfer set for %d", token.Token),
+			"err", lastErr,
+		)
+	}
+	return ret, lastErr == nil
+}
+
+// joinIncrementalTransfer will attempt to incrementally obtain chunks from
+// neighboring lifecyclers that contain data for token ranges they will
+// no longer receive writes for.
+func (i *Lifecycler) joinIncrementalTransfer(ctx context.Context) error {
+	// Make sure that we set all tokens to ACTIVE, even
+	// when we fail.
+	defer func() {
+		i.setTokens(i.getTransitioningTokens())
+		i.setTokensState(ACTIVE)
+		i.changeState(ctx, ACTIVE)
+		i.updateConsul(ctx)
+	}()
+
+	r, err := i.getDenormalisedRing(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to read ring: %v", err)
+	}
+
+	replicationFactor := i.cfg.RingConfig.ReplicationFactor
+	if active := r.FindIngestersByState(ACTIVE); len(active) < replicationFactor {
+		if len(active) == 0 {
+			return fmt.Errorf("no ingesters to request data from")
+		}
+
+		// Request an entire copy of all chunks from the first replica.
+		fullRange := []TokenRange{{0, math.MaxUint32}}
+		err := i.incTransferer.RequestChunkRanges(ctx, fullRange, active[0].Addr, false)
+		if err != nil {
+			level.Error(util.Logger).Log("msg",
+				fmt.Sprintf("failed to request copy of data from %s", active[0].Addr))
+		}
+
+		return nil
+	}
+
+	var pendingUnblocks []struct {
+		addr string
+		rgs  []TokenRange
+	}
+
+	var wg sync.WaitGroup
+
+	for tokidx, token := range i.transitioningTokens {
+		// For findTransferWorkload to find the full list of tokens to transfer,
+		// we must first set the token to JOINING locally and add it into our
+		// copy of the ring.
+		i.stateMtx.Lock()
+		token.State = JOINING
+		i.transitioningTokens[tokidx] = token
+		i.stateMtx.Unlock()
+
+		r.AddToken(i.ID, token, false)
+
+		workload, _ := i.findTransferWorkload(r, token)
+
+		for addr, ranges := range workload {
+			wg.Add(1)
+			go func(addr string, ranges []TokenRange) {
+				err := i.incTransferer.BlockRanges(ctx, ranges, addr)
+				if err == nil {
+					err = i.incTransferer.RequestChunkRanges(ctx, ranges, addr, true)
+				}
+
+				if err != nil {
+					level.Error(util.Logger).Log(
+						"msg", "failed to request chunks",
+						"target_addr", addr,
+						"ranges", PrintableRanges(ranges),
+						"err", err,
+					)
+				}
+
+				wg.Done()
```

Comment:
It's a good practice to call `wg.Done()` via `defer` at the beginning of the function (multiple places in this PR)